### PR TITLE
Simplify port API and add some operator tasks

### PR DESCRIPTION
**Pull request by jbeezley**
This goes on top of #65.  The main purpose of this PR is to make it easier to bind data to task ports, but it also adds some basic operators for data types that support them.

```python
from gaia.core import operators

add1 = operators.Add()
add2 = operators.Add()
multiply = operators.Multiply()

add1.set_input(1, 2)
add2.set_input(3, 4)
multiply.set_input(add1.get_output(), add2.get_output())

multiply.get_output_data() # returns 21
```

See the docstring for [`task.set_input`](https://github.com/Kitware/romanesco/blob/simple-operators/gaia/core/task.py#L50) for details.

### CODE REVIEW

**Comment by aashish24** [comment]
Thanks @jbeezley I will have it reviewed sometime today..

File: CMakeLists.txt [hunk 1]
```diff
@@ -16,6 +16,7 @@ if(PYTHON_COVERAGE)
 endif()
 
 if(PYTHON_STYLE_TESTS)
+  find_program(FLAKE8_EXECUTABLE flake8)
   find_program(PEP8_EXECUTABLE pep8)
```
Comment (aashish24) [inline CMakeLists.txt]: do we still need pep8?

**Comment by jeffbaumes** [comment]
I think your `operator` sounds more like a reducer in data science terms - should the name be changed? I feel like if it is meant to do 1, 2, 3 or so inputs it's an operator, but if it is meant to handle hundreds of inputs it becomes better to call it a reducer.

**Comment by aashish24** [comment]
@jeffbaumes not sure if I like reducer as much but I will okay if we call it reducer. any other suggestions?

File: gaia/core/task.py [hunk 2]
```diff
@@ -121,11 +169,9 @@ def run(self, *arg, **kw):
         the input ports are all connected and raises an error if they
         aren't.
         """
-        for name in self.input_ports:
-            itask = self.get_input_task(name)
-            if itask is None:
-                raise Exception("Input port '{0}' not connected.".format(name))
-            self._input_data[name] = itask.get_output_data(name)
+        self.dirty = False
```
Comment (aashish24) [inline gaia/core/task.py]: We may have to move to start using timestamps but that would be in a separate PR

File: gaia/core/task.py [hunk 3]
```diff
@@ -166,16 +212,16 @@ class Source(Task):
...
             """Generated source task."""
 
-            output_ports = {'': SourceOutput}
+            output_ports = {'0': SourceOutput}
 
-            def __init__(self, *arg, **kw):
-                """Initialize a source task."""
-                super(Source, self).__init__(*arg, **kw)
-                self._output_data[''] = data
+            def get_input_data(self, name='0'):
```
Comment (aashish24) [inline gaia/core/task.py]: should this be private (with underscore) since we don't have set_input_data. Looks like it will be used internally by the program and not by the user?

File: gaia/pandas/geopandas_plot.py [hunk 4]
```diff
@@ -10,17 +10,16 @@ class GeopandasPlot(Task):
...
     input_ports = {
-        '': Task.make_input_port(GeoDataFrame)
+        '0': Task.make_input_port(GeoDataFrame)
     }
 
     def run(self, *args, **kw):
         """Plot data using geopandas."""
         super(GeopandasPlot, self).run(*args, **kw)
-        data = self._input_data['']
+        data = self.get_input_data()
```
Comment (aashish24) [inline gaia/pandas/geopandas_plot.py]: +1

File: gaia/pandas/geopandas_reader.py [hunk 5]
```diff
@@ -12,19 +12,18 @@ class GeopandasReader(Task):
...
-        self._output_data[''] = None
+        super(GeopandasReader, self)._reset(*args)
+        self._output_data['0'] = None
 
     def run(self, *args, **kw):
         """Read and cache file data using geopandas."""
         super(GeopandasReader, self).run(*args, **kw)
-        self._output_data[''] = read_file(self.file_name)
-        self.dirty = False
+        self._output_data['0'] = read_file(self.file_name)
```
Comment (aashish24) [inline gaia/pandas/geopandas_reader.py]: How about adding an API set_output_data()? similar to get_input_data.

File: tests/CMakeLists.txt [hunk 6]
```diff
@@ -34,8 +34,9 @@ if(PYTHON_COVERAGE)
 endif()
 
 if(PYTHON_STYLE_TESTS)
-  add_python_style_test(pep8_style_core "${PROJECT_SOURCE_DIR}/romanesco")
-  add_python_style_test(pep8_style_core_tests "${PROJECT_SOURCE_DIR}/tests")
+  add_python_pep8_test(pep8_style_core "${PROJECT_SOURCE_DIR}/romanesco")
```
Comment (aashish24) [inline tests/CMakeLists.txt]: I see that we are still using pep8.

File: tests/geopandas_test.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        writer.run()
+
+        self.assertTrue(os.path.isfile(fname))
+
+
+class GeopandasPlotTest(TestCase):
+
+    """Test the geopandas plotter."""
+
+    def setUp(self):  # noqa
```
Comment (aashish24) [inline tests/geopandas_test.py]: just curious: looks like other function names use underscore style vs this one and some others uses camel case style.

**Comment by aashish24** [comment]
I am getting style tests failing on my system: 

    Start 2: flake8_style_gaia

2: Test command: /usr/local/bin/flake8 "/home/chaudhary/tools/romanesco/src.git/gaia"
2: Test timeout computed to be: 1500
2: Traceback (most recent call last):
2:   File "/usr/local/bin/flake8", line 11, in <module>
2:     sys.exit(main())
2:   File "/usr/local/lib/python2.7/dist-packages/flake8/main.py", line 24, in main
2:     flake8_style = get_style_guide(parse_argv=True, config_file=DEFAULT_CONFIG)
2:   File "/usr/local/lib/python2.7/dist-packages/flake8/engine.py", line 110, in get_style_guide
2:     kwargs['parser'], options_hooks = get_parser()
2:   File "/usr/local/lib/python2.7/dist-packages/flake8/engine.py", line 48, in get_parser
2:     (extensions, parser_hooks, options_hooks, ignored) = _register_extensions()
2:   File "/usr/local/lib/python2.7/dist-packages/flake8/engine.py", line 32, in _register_extensions
2:     checker = entry.load()
2:   File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 2087, in load
2:     if require: self.require(env, installer)
2:   File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 2100, in require
2:     working_set.resolve(self.dist.requires(self.extras),env,installer)))
2:   File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 632, in resolve
2:     raise VersionConflict(dist,req) # XXX put more info here
2: pkg_resources.VersionConflict: (pep8 1.6.2 (/usr/local/lib/python2.7/dist-packages), Requirement.parse('pep8>=1.5.7,<1.6'))

File: tests/geopandas_test.py (hunk 7 above)
Comment (jbeezley) [inline tests/geopandas_test.py]: That is python's unittest API.  The `noqa` is there because pep8 doesn't like it.  (Yeah, python's own API doesn't conform to pep8 standards...)

Comment (aashish24) [inline tests/geopandas_test.py]: Ah..yes sorry forgot about that. thanks for the info.

File: tests/CMakeLists.txt (hunk 6 above)
Comment (aashish24) [inline tests/CMakeLists.txt]: I meant still using pep8 directly

**Comment by jbeezley** [comment]
Your pep8/flake8 install is broken.  I would try a `pip install -U -r requirements-dev.txt`.  Also, your python install looks unusual, both `/usr/local/lib/python2.7/dist-packages` and `/usr/lib/python2.7/dist-packages`.

File: tests/CMakeLists.txt (hunk 6 above)
Comment (jbeezley) [inline tests/CMakeLists.txt]: The romanesco code uses `pep8`.  We will have to work on it to get it to pass `flake8`.

**Comment by aashish24** [comment]
pip install -U -r requirements-dev.txt fixed my problem.  pip installs packages under /usr/local/lib and rest of the libs are under /usr/lib. Were you expecting something else?

**Comment by jeffbaumes** [comment]
@aashish24 I just see this could be used for many other more complex things than add or multiply which might not really be thought of as operators. For example, you could use this to combine many histograms into a resulting histogram. I agree though that `reducer` does not sound quite as pleasant.

**Comment by jbeezley** [comment]
@jeffbaumes Originally, my plan was for much more fine grained tasks along with control components providing looping and other control logic that could implement parallelism.  From our discussions, that plan was too ambitious for the moment.  Because these very simple tasks provide a good way to play around with and test the API, I decided to keep them around.

File: tests/geopandas_test.py (hunk 7 above)
Comment (zachmullen) [inline tests/geopandas_test.py]: from cStringIO import StringIO

:wink:

**Comment by jeffbaumes** [comment]
@jbeezley Ok I'm totally ok with `operator` if these are just play examples for now. But keep in mind that code has a funny way of sticking around and growing :seedling:

**Comment by aashish24** [comment]
@jeffbaumes yes that's what the idea is. I am sure we will have to make another pass on some of these things. I liked your :seedling: --> :palm_tree:

**Comment by jeffbaumes** [comment]
Aashish do you realize this branch was never merged? I think you may have closed the PR accidentally earlier today.

**Comment by aashish24** [comment]
thanks -:) I was talking and working at the same time :smile: now its merged
